### PR TITLE
Add multi-module code coverage report using configurations

### PR DESCRIPTION
**Pull request by bmuskalla**

### CODE REVIEW

**Comment by bmuskalla** [comment]
Please ignore the kotlin part for now; we should have a quick discussion how to move forward with this. I really like how the approach works out on the build file level but it is a lot of code in buildSrc that people would need to copy around.

**Comment by bmuskalla** [comment]
Fully functional sample project is here: https://github.com/bmuskalla/gradle-coverage-configurations

File: subprojects/docs/src/samples/templates/jvm-coverage-configurations/buildSrc/src/main/java/com/example/jacoco/report/JacocoCollectCoveragePlugin.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        final JavaPluginConvention plugin = project.getConvention().getPlugin(JavaPluginConvention.class);
+        final ConfigurationPublications outgoing = coverageSourcePathConfiguration.getOutgoing();
+        plugin.getSourceSets().findByName("main").getAllJava().forEach(outgoing::artifact);
+        plugin.getSourceSets().findByName("test").getAllJava().forEach(outgoing::artifact);
+    }
+
+    private void configureTestToPublishCoverageData(Project project) {
+        final Task testTask = project.getTasks().getByName("test");
+        final File destinationFile = testTask.getExtensions().findByType(JacocoTaskExtension.class).getDestinationFile();
+        project.getArtifacts().add("coverageData", destinationFile, a -> a.builtBy(testTask));
```
Comment (vlsi) [inline subprojects/docs/src/samples/templates/jvm-coverage-configurations/buildSrc/src/main/java/com/example/jacoco/report/JacocoCollectCoveragePlugin.java]: 1) `JacocoTaskExtension` might be not present
2) `JacocoTaskExtension` might be applied earlier than `jacoco`, so in theory it should probably be inside `afterEvaluate` block (or so)
3) There might be multiple `test` tasks
4) Configuration avoidance would be a plus

File: subprojects/docs/src/samples/templates/jvm-coverage-configurations/buildSrc/src/main/java/com/example/jacoco/report/JacocoReportConfigurations.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.Category;
+import org.gradle.api.attributes.DocsType;
+import org.gradle.api.attributes.Usage;
+import org.gradle.api.model.ObjectFactory;
+
+public class JacocoReportConfigurations {
```
Comment (big-guy) [inline subprojects/docs/src/samples/templates/jvm-coverage-configurations/buildSrc/src/main/java/com/example/jacoco/report/JacocoReportConfigurations.java]: I think it's probably more straightforward to either have separate clear methods that create configurations for the producer and consumer or to put the differences in the plugins themselves.  I think having a flag to switch between producer/consumer oversimplifies how a producer and consumer can be different.

File: subprojects/docs/src/samples/jvm-coverage-configurations/README.adoc [hunk 3]
```diff
@@ -0,0 +1,32 @@
+NOTE: You can open this sample inside an IDE using the https://www.jetbrains.com/help/idea/gradle.html#gradle_import_project_start[IntelliJ native importer] or https://projects.eclipse.org/projects/tools.buildship[Eclipse Buildship].
+
+This sample shows how to use dependencies within a JVM multi-project build in Gradle. This is the same pattern used by all JVM languages (Java, Groovy, Scala).
```
Comment (big-guy) [inline subprojects/docs/src/samples/jvm-coverage-configurations/README.adoc]: This looks like it's a copy from the multi-project build sample.  I think we need some "theory of operation" with the custom plugin here so people know what to expect.

File: subprojects/docs/src/samples/templates/jvm-coverage-configurations/buildSrc/src/main/java/com/example/jacoco/report/JacocoReportAllPlugin.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public void apply(Project project) {
+        final JacocoReportConfigurations configurations = JacocoReportConfigurations.forConsumers(project);
+        configurations.createRootConfiguration();
+        createReportCoverageTask(project, configurations);
+    }
+
+    private void createReportCoverageTask(Project project, JacocoReportConfigurations configurations) {
+        project.getTasks().register("codeCoverageReport", JacocoReport.class, reportTask -> {
+            reportTask.additionalClassDirs(configurations.createCoverageRuntimeConfiguration());
```
Comment (big-guy) [inline subprojects/docs/src/samples/templates/jvm-coverage-configurations/buildSrc/src/main/java/com/example/jacoco/report/JacocoReportAllPlugin.java]: can we also set group/description?

**Comment by bmuskalla** [comment]
Thanks for all the valuable feedback folks! I'm really sorry, this PR was not intended to be reviewed (that's why I submitted it as draft), it was to start a discussion between some folks if we really want to publish the configuration approach as a sample. For now, we decided to not further push this sample but rather clarify the usage of the `JacocoReport` API and its hiccups (and advantages over configurations). I'll soon push another sample using those APIs.

**Pull request closed without merging**